source/rpt/rpt_fem_reconstruction.cc

```diff
@@ -0,0 +1,1080 @@
+#include <deal.II/base/multithread_info.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/work_stream.h>
+
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_in.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/manifold_lib.h>
+
+#include <deal.II/lac/lapack_full_matrix.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+
+#include <deal.II/numerics/data_out.h>
+
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/archive/text_oarchive.hpp>
+
+#include <rpt/particle_detector_interactions.h>
+#include <rpt/rpt_fem_reconstruction.h>
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+using namespace dealii;
+
+
+template <int dim>
+RPTFEMReconstruction<dim>::AssemblyScratchData::AssemblyScratchData(
+  const FiniteElement<dim> &fe,
+  const unsigned int        no_detector)
+  : fe_values(fe,
+              QGaussSimplex<dim>(fe.degree + 1),
+              update_values | update_quadrature_points | update_JxW_values)
+  , detector_id(no_detector)
+{}
+
+template <int dim>
+RPTFEMReconstruction<dim>::AssemblyScratchData::AssemblyScratchData(
+  const AssemblyScratchData &scratch)
+  : fe_values(scratch.fe_values.get_fe(),
+              scratch.fe_values.get_quadrature(),
+              update_values | update_quadrature_points | update_JxW_values)
+  , detector_id(scratch.detector_id)
+{}
+
+
+template <int dim>
+void
+RPTFEMReconstruction<dim>::setup_triangulation()
+{
+  TimerOutput::Scope t(computing_timer, "setting_up_grid");
+
+  triangulation.clear();
+
+  if (rpt_parameters.fem_reconstruction_param.mesh_type ==
+      Parameters::RPTFEMReconstructionParameters::FEMMeshType::gmsh)
+    {
+      GridIn<dim> grid_in;
+      grid_in.attach_triangulation(triangulation);
+      std::ifstream input_file(
+        rpt_parameters.fem_reconstruction_param.mesh_file);
+      grid_in.read_msh(input_file);
+
+      const CylindricalManifold<dim> manifold(2);
+      triangulation.set_all_manifold_ids(0);
+      triangulation.set_manifold(0, manifold);
+    }
```

Comment:
Would add more detail about the setup (explain that the gmsh have to be a cylinder with z as height axis). Also, is there a check about that if the mesh is not a cylinder?